src/ESPAsyncTCP.cpp

```diff
@@ -395,7 +395,7 @@ err_t AsyncClient::_recv(tcp_pcb* pcb, pbuf* pb, err_t err) {
         ASYNC_TCP_DEBUG("_recv err: %d\n", read_bytes);
         _close();
       }
-      return read_bytes;
+      //return read_bytes;
```

Comment:
Shouldn't we return error instead of saying ERR_OK?

Reply:
Not really:
1. if tcp_ssl_read() returns SSL_CLOSE_NOTIFY
    - The pbuf has been freed in tcp_ssl_read(), so LwIP should not try to further maintain it, and
    - Freeing pbuf in tcp_ssl_read() is a good choice because by definition of this signal, we are closing the connection anyway.
2. if tcp_ssl_read() returns other error code
    - The block above the return is closing the connection anyway, so we don't care the left over data in pbuf
    - There are two code path in tcp_ssl_read(), one frees the pbuf, one does not (handshake error)
        - For the code path that does not free pbuf, a reference is left in fd_data->tcp_pbuf, which will be freed when tcp_ssl_free() is called later.
    - Therefore, it is still necessart and correct to let LwIP not further maintain the pbuf.

So, in short, when we run into read_bytes < 0 situation, we will be closing the connection one way or another, and it is best to let LwIP give up ownership of the current pbuf.

And the easiest way to let LwIP to what we want is to return ERR_OK.

Reply:
From a different way to look at it, SSL belongs to "upper" layer and TCP belongs to "lower" layer.

SSL_CLOSE_NOTIFY is an upper layer signal, it may mean some abnormal situation, only to the SSL layer.

To the lower TCP layer, everything is normal -- the TCP layer successfully delivered the close notification signal to the SSL layer, hence ERR_OK.